---
title: "Decontamination of ambient RNA in single-cell RNA-seq with DecontX on 10X PBMC4k data" 
author: "Shiyi Yang, Sean Corbett, Yusuke Koga, Zhe Wang, W. Evan Johnson, Masanao Yajima, Joshua D. Campbell"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Estimate and remove cross-contamination from ambient RNA for scRNA-seq data with DecontX}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r, eval=TRUE, message=FALSE}
library(celda)
```

DecontX can take either `SingleCellExperiment` object from [SingleCellExperiment package] (https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html) or a count matrix in the format of rows being genesand columns being cells. When using SingleCellExperiment object, DecontX is using the counts matrix extracted from the SingleCellExperiment object. 

# TODO: Load PBMC4k data from 10X CellRanger output
# Load PBMC4k data from 10X
```{r, eval=TRUE, message=FALSE}
library(TENxPBMCData)  # selected single cell RNAseq datasets from 10X Genomics
pbmc <- TENxPBMCData("pbmc4k")
colnames(pbmc) <- paste(pbmc$Sample, pbmc$Barcode, sep="_") # Rename cells
```

Take a look at the count matrix in the `pbmc` singleCellExperiment object via `assay` function from [SummarizedExperiment] (https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html). There are total 33694 genes and 4340 cells in this PBMC data.
```{r}
library(SummarizedExperiment)
pbmc_counts = SummarizedExperiment::assay(pbmc, i = "counts")
print(pbmc_counts)
```

Among the 33694 genes, many of then are empty genes (i.e., not expressed by any cell), and many of the genes are likely to be noise due to extremely low expression among these 4340 cells.
We filter out genes that are likely to be noise. We normally keep genes that have expressed at least 3 counts in at least 3 cells.
After filtering, we have 4529 genes left.
```{r}
#pbmc_filterg = pbmc[rowSums(pbmc_counts > 2) > 2, ] 
pbmc_filterg = pbmc[rowSums(pbmc_counts >= 3) >= 3, ] # Filtering can be applied directly on SingleCellExperiment object 
print(pbmc_filterg)
```



# Estimate and remove contaminated counts using DecontX (2 scenarios -- with and without cell types specified)
## Use DecontX for contamination estimation when there is cell type specified for each individual cells (Here we use celda_CG -- a Bayesian bi-clustering method on single cell RNAseq data -- to get cell clusters as well as gene modules, which enhances cell clustering. You can substitute `celda_CG` with any clustering method you have faith in. DecontX takes the expression count matrix (gene by cell count matrix) and a vector of the same length as number of cells in the count matrix specifying cell cluster. 
```{r}
pbmc_filterg_counts = SummarizedExperiment::assay(pbmc_filterg, "counts")
pbmc_filterg_counts = as.matrix(pbmc_filterg_counts) # Change obejct type as basic matrix to pass into celda::celda_CG
celda_res = celda_CG(counts = pbmc_filterg_counts, K = 19, L = 150) # specify number of cell clusters being K=19, number of gnee modules being L=150
```
P.s: we have analyzed this 4K PBMC data before, so we don't bother with the parameters choice for the bi-clustering method `celda_CG` here. For a detailed toturial of usage on `celda`, it is availabel [here] (celda-vignettes url) 



Pass expression (gene by cell) count matrix and the cluster labels from `celda_CG` to decontX, it will return estimated results. Corresponding to the 2 types of input, output result is either in a `list`, or a SingleCellExperiment object with the result list added back to its `metadata` slot (?? Am I understanding the structure right? Is metadata a "slot" of the sce-object??). When `decontX` takes a SingleCellExperiment object, we can use function `metadata` to extract the result list and use `str` function to look at the 4 attributes ("runParams", "estimates", "contamination", "z") in it.
```{r}
cell_cluster = celda_res@clusters$z 
decontx_res_sce = decontX( x = pbmc_filterg, z = cell_cluster) 
print(decontx_res_sce)
#decontx_res = decontX( x = pbmc_filterg_counts, z = cell_cluster) ## This works too. Difference is that the output is a list rathen than being a SingleCellExperiment object
decontx_res = metadata(decontx_res_sce)$decontX
str(decontx_res)
```


## Interpretation of DecontX results
There are 4 attributes of DecontX results: 
(1) runParams: parameters used for estimating contamination
(2) estimates: parameters, decontaminated counts as well as other intermediate results estimated during contamination estimation
(3) contamination: cell level contamination
(4) z: cell cluster label 


# Check DecontX decontaminated count to see if gene expressions are cleaner by looking at marker genes' expression
## First we have to identify cell types, which is one step further from cell clustering, as we do not really know which cluster is corresponding to what cell type and multiple cell clusters can also be from the same cell type. We are looking at the 4 major cell types (T-cell, B-cell, NK-cell and monocytes) in this PBMC data.
Combining both the clustering results from `celda_CG`, visualization on a reduced 2-dimention space using Tsne (T-stochastic Neighbor Embedding), and marker genes, we relate cell clusters (more specifically the cells in them) to cell types. The marker genes used to identify these 4 major cell types are CD3D and CD3E (T-cell markers), MS4A1, CD79A and CD79B and (B-cell markers), GNLY (NK-cell marker), LYZ, S100A8, and S100A9 (monocytes markers). There are also other cell types (such as dendritic cells, plasma cells, etc.), we don't include these cell types in our demonstration here and we point you to [here] (celda-vignettes url on PBMC4k data) for cell type analysis using `celda`.
```{r}
celdaTsne.ori = celdaTsne( counts = pbmc_filterg_counts , celdaMod = celda_res ) 
pltTsne.ori = plotDimReduceCluster( dim1 = celdaTsne.ori[,1], dim2 = celdaTsne.ori[, 2], cluster = cell_cluster, labelClusters = T)
pltTsne.ori
```
violin plot of marker genes 
marker genes are those used in manuscript

We use violin plot to look at the marker gene expression levels in each cluster identified before by `celda_CG`. Those gene names are stored in the "rowData" of the singleCellExperiment object. On the other hand, the rownames of the expression count matrix is using gene ID. We have to match marker genes' name to the gene ID to locate them in the expression count matrix.
```{r}
markerGenes = do.call(what = rbind, args = list(data.frame("celltype" = "Tcell", "markergenes"=c("CD3D", "CD3E")), 
		                                data.frame("celltype" = "Bcell", "markergenes"=c("MS4A1", "CD79A", "CD79B")),
			                        data.frame("celltype" = "NKcell", "markergenes"=c("GNLY")),
			                        data.frame("celltype" = "monocytes", "markergenes"=c("LYZ", "S100A8", "S100A9"))))
print(markerGenes)
head(rowData(pbmc_filterg))
markerGenes_wID = merge( x = markerGenes, y = rowData(pbmc_filterg), by.x = "markergenes", by.y = "Symbol_TENx")
markerGenes_wID[, "feature"] = paste( markerGenes_wID[, "ENSEMBL_ID"], markerGenes_wID[, "markergenes"], sep = "_")
print(markerGenes_wID)

pbmc_filterg_counts_rename = pbmc_filterg_counts
rownames(pbmc_filterg_counts_rename) = paste(rownames(pbmc_filterg), rowData(pbmc_filterg)[, "Symbol"], sep = "_")
pltViolin.ori = violinPlot(counts = pbmc_filterg_counts_rename, celdaMod=celda_res, features = markerGenes_wID[, "feature"])
pltViolin.ori
```





## Use DecontX for contamination estimation when there is no cell type specified for the cells





