---
title: "Decontamination of ambient RNA in single-cell RNA-seq with DecontX on 10X PBMC4k data" 
author: "Shiyi Yang, Sean Corbett, Yusuke Koga, Zhe Wang, W. Evan Johnson, Masanao Yajima, Joshua D. Campbell"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Estimate and remove cross-contamination from ambient RNA for scRNA-seq data with DecontX}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r, eval=TRUE, message=FALSE}
library(celda)
```

DecontX can take either `SingleCellExperiment` object from [SingleCellExperiment package] (https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html) or a count matrix in the format of rows being genesand columns being cells.  

# TODO: Load PBMC4k data from 10X CellRanger output
# Load PBMC4k data from 10X
```{r, eval=TRUE, message=FALSE}
library(TENxPBMCData)  # selected single cell RNAseq datasets from 10X Genomics
pbmc <- TENxPBMCData("pbmc4k")
colnames(pbmc) <- paste(pbmc$Sample, pbmc$Barcode, sep="_")
```

Take a look at the count matrix in the `pbmc` singleCellExperiment object via `assay` function from [SummarizedExperiment] (https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html)
```{r}
library(SummarizedExperiment)
pbmc_counts = SummarizedExperiment::assay(pbmc, i = "counts")
```


Filter out genes that are likely to be noise. We normally choose genes that have expressed at least 3 counts in at least 3 cells.
```{r}

```



# Estimate and remove contaminated counts using DecontX (2 scenarios -- with and without cell types specified)
## Use DecontX for contamination estimation when there is no cell type specified for the cells
We use celda_CG -- a Bayesian bi-clustering method on single cell RNAseq data -- to get cell clusters (as well as gene modules, which enhances cell type identification). 
```{r}

```




## Use DecontX for contamination estimation when there is cell type specified for each individual cells
